src/ssl/PeekingPeerConnector.cc

```diff
@@ -128,17 +129,12 @@ Ssl::PeekingPeerConnector::checkForPeekAndSpliceMatched(const Ssl::BumpMode acti
 Ssl::BumpMode
 Ssl::PeekingPeerConnector::checkForPeekAndSpliceGuess() const
 {
-    if (const ConnStateData *csd = request->clientConnectionManager.valid()) {
-        const Ssl::BumpMode currentMode = csd->sslBumpMode;
-        if (currentMode == Ssl::bumpStare) {
-            debugs(83,5, "default to bumping after staring");
-            return Ssl::bumpBump;
-        }
-        debugs(83,5, "default to splicing after " << currentMode);
-    } else {
-        debugs(83,3, "default to splicing due to missing info");
+    if (currentBumpMode == Ssl::bumpStare) {
+        debugs(83,5, "default to bumping after staring");
+        return Ssl::bumpBump;
     }
-
+    Must(currentBumpMode == Ssl::bumpPeek);
+    debugs(83,5, "default to splicing after " << currentBumpMode);
     return Ssl::bumpSplice;
```

Comment:
> 1. Why currentBumpMode cannot be bumpServerFirst here? Your earlier [comment](https://github.com/measurement-factory/squid/pull/139#issuecomment-959828602) appears to imply that it may be (because when doing server-first bumping, we skip step2 and bumpModeAtStep(tlsBump2) returns step1's mode).

We are checking acls at Bumping step3 only if at step2 we decide to peek or stare. If at step2 we decided to bump then this is a final decision we are going to just TLS-connect  with the remote and bump.

If you check the Ssl::PeekingPeerConnector::noteWantWrite method you will see that we are calling checkForPeekAndSplice() method (and thus the checkForPeekAndSpliceGuess) only if we are in Peek or Stare modes.


> 
>     2. Why currentBumpMode cannot be bumpNone here? PeekingPeerConnector constructor leaves currentBumpMode with that "default" value in some cases.

We never enters acls check sequence at step3 if the decision on step2 was not peek or stare.
The currentBumpMode is set to bumpNone in PeekingPeerConnector  constructor when the client gone,  as I can see PeekingPeerConenctor will just initiate a TLS connection with the remote in this case (no acls at step3 will be checked).

Probably in PeekingPeerConnector constructor we can throw if client (ConnStateData object) is gone to not allow PeekingPeerConnector start at all in this case.

> 
>     3. Why currentBumpMode cannot be bumpBump here? The "Ssl::bumpServerFirst or Ssl::bumpBump bumping modes" comment added by this PR to initialize() appears to imply that it may be.

In PeekingPeerConnector::initialize it can be "Ssl::bumpServerFirst or Ssl::bumpBump". But if in initialize the currentBumpMode has these values then the PeekingPeerConnector just connects to the remote server to bump, without check the acls at step3.

Reply:
> > 1. Why currentBumpMode cannot be bumpServerFirst here?

I think I was able to refactor your answer to get to the bottom of this. Here is the refactored answer for the record:

* checkForPeekAndSpliceGuess() is currently only called from checkForPeekAndSpliceDone().
* checkForPeekAndSpliceDone() is currently only called from checkForPeekAndSplice().
* checkForPeekAndSplice() is currently only called when currentBumpMode is bumpPeek or bumpStare.
* currentBumpMode does not change during PeekingPeerConnector lifetime.

Thus, the Must() in question will not be violated in the current code. None of the above assumptions are explicit, so it is not unlikely that at least one of them will eventually change. We can expose our assumptions by adding a source code comment or even changing all the "checkForPeekAndSplice" prefixes to something that makes most of our assumptions explicit (e.g., "postPeekingOrStaring"), but it is also possible that the whole class needs to be split so that we do not have to deal with multiple assumption-heavy logical paths inside one class. Such a split would be outside this PR scope IMO.

> > 2. Why currentBumpMode cannot be bumpNone here? PeekingPeerConnector constructor leaves currentBumpMode with that "default" value in some cases.

> We never enters acls check sequence at step3 if the decision on step2 was not peek or stare.

Right, same answer as Question 1.

> The currentBumpMode is set to bumpNone in PeekingPeerConnector constructor when the client gone, as I can see PeekingPeerConenctor will just initiate a TLS connection with the remote in this case

4. Will the server connection be bumped or spliced in this gone-client case?

Splicing with the gone client connection is usually pointless. Bumping can be useful if we can cache the HTTP response and/or share it with other clients.

> Probably in PeekingPeerConnector constructor we can throw if client (ConnStateData object) is gone to not allow PeekingPeerConnector start at all in this case.

Our constructor comment says "we cannot check the step, but must carry on" so there was a reason why we thought we have to continue. One reason could be that quick_abort or similar settings allow Squid to cache the server response even after the client is long gone. Perhaps we did not want to introduce a special case where that kind of gone-client fetching/caching cannot happen?
 
> > 3. Why currentBumpMode cannot be bumpBump here?

> In PeekingPeerConnector::initialize it can be "Ssl::bumpServerFirst or Ssl::bumpBump". But if in initialize the currentBumpMode has these values then the PeekingPeerConnector just connects to the remote server to bump, without check the acls at step3.

Yes, same answer as Question 1.

I consider all three original questioned answered. Thank you. Please see the new question 4 above.

Reply:
> Will the server connection be bumped or spliced in this gone-client case?
> 
> Splicing with the gone client connection is usually pointless. Bumping can be useful if we can cache the HTTP response and/or share it with other clients.

We are just TLS connecting to the server, I do not think that the TLS connection will survive after the connection returned to the FwdState object.  The HTTP request will never reach the server.

Reply:
>  The HTTP request will never reach the server.

I think you are implying that there is no HTTP request that FwdState can forward on the bumped connection. That sounds correct today, but I am not 100% sure that it is a fundamental limitation/principle. I wonder whether something like TLS 0-RTT (when supported) might give Squid a usable HTTP request at server bumping time even if the client is gone.

Perhaps more importantly, it feels wrong that PeekingPeerConnector (a class that has nothing to do with HTTP requests and, essentially, with the client itself) has to worry about client existence! I suspect that correctly written code should never ask the "Is there a client?" question inside PeekingPeerConnector. The question itself should be out of PeekingPeerConnector scope. We should not be thinking about the best reaction to the "no" answer; we should not be asking the question instead.

Today, the code surrounding PeekingPeerConnector makes that wrong question _necessary_ because the client object stores some state that we need. That storage is misplaced, and we should fix that, but this PR is not a good place for _that_ fix.

I consider this change request addressed. We do not know the answer to Question 4, but perhaps there is no easy answer because the client can disappear at virtually any time and there are several gone client checks in PeekingPeerConnector code. The set of possible outcomes is probably to varied to describe in a simple answer.